Web Extensions CodeGenerator needs to handle objects better.

xeenon opened the pull request:
#### 60e0f080e0bd7519786d780c6890ed2a11ea6748
<pre>
Web Extensions CodeGenerator needs to handle objects better.
<a href="https://webkit.org/b/262496">https://webkit.org/b/262496</a>
rdar://problem/116359645

Reviewed by Brent Fulgham.

The object checks were too aggressively converting things to an NSDictionary, when only basic Objects
should be converted, and everything else should be a JSValue. This is needed for the action APIs which
allow ImageData objects, and was requiring the action code to manually deal with JSValue for all input.

* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::toNSObject): Change logic for when to use JSValue by checking _isDictionary and other native types.
(WebKit::toNSDictionary): Recursively use toNSDictionary() for dictionary items.
(-[JSValue _isDictionary]): Added.
(-[JSValue _isRegularExpression]): Do a proper type check.
(-[JSValue _isThenable]): Use _isFunction for clarity.
(-[JSValue _awaitThenableResolutionWithCompletionHandler:]): Use _isThenable for clarity.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::toNSArray): Added wrapper that ASSERTs containingObjectsOfClass and defaults to NSObject type.
(WebKit::toJSValue): Add null check for value and ASSERT for context.
(WebKit::isDictionary): Added. C version of -[JSValue _isDictionary].
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile): Use $defaultEarlyReturnValue when needed, this was incorrect before
when $defaultReturnValue can also still return the promise result during an exception.
(_javaScriptTypeCondition): Use isDictionary() and JSValueIsArray for dictionary and array checks.
(_platformTypeConstructor): Support NullAllowed, which is needed for action. Use toNSArray.

Canonical link: <a href="https://commits.webkit.org/268747@main">https://commits.webkit.org/268747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52a70191f0a2f443e3b4bd329b2d8b53fc0a5739

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20545 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20969 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21618 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19163 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24224 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21146 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20547 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20765 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17860 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23299 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17781 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24948 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22886 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19431 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16509 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18648 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22985 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2539 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19250 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
Looks good, though I'm trusting your judgement on the Function/Thenable/Dictionary logic details. :-)

xeenon committed:
Web Extensions CodeGenerator needs to handle objects better.
https://webkit.org/b/262496
rdar://problem/116359645

Reviewed by Brent Fulgham.

The object checks were too aggressively converting things to an NSDictionary, when only basic Objects
should be converted, and everything else should be a JSValue. This is needed for the action APIs which
allow ImageData objects, and was requiring the action code to manually deal with JSValue for all input.

* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::toNSObject): Change logic for when to use JSValue by checking _isDictionary and other native types.
(WebKit::toNSDictionary): Recursively use toNSDictionary() for dictionary items.
(-[JSValue _isDictionary]): Added.
(-[JSValue _isRegularExpression]): Do a proper type check.
(-[JSValue _isThenable]): Use _isFunction for clarity.
(-[JSValue _awaitThenableResolutionWithCompletionHandler:]): Use _isThenable for clarity.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::toNSArray): Added wrapper that ASSERTs containingObjectsOfClass and defaults to NSObject type.
(WebKit::toJSValue): Add null check for value and ASSERT for context.
(WebKit::isDictionary): Added. C version of -[JSValue _isDictionary].
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile): Use $defaultEarlyReturnValue when needed, this was incorrect before
when $defaultReturnValue can also still return the promise result during an exception.
(_javaScriptTypeCondition): Use isDictionary() and JSValueIsArray for dictionary and array checks.
(_platformTypeConstructor): Support NullAllowed, which is needed for action. Use toNSArray.

Canonical link: https://commits.webkit.org/268747@main
